modules/ref_fb_module/src/fft_fb_impl.cpp

```diff
@@ -0,0 +1,234 @@
+
+        if (inputDomainDataDescriptor.getSampleType() != SampleType::Int64 && inputDomainDataDescriptor.getSampleType() != SampleType::UInt64)
+        {
+            throw std::runtime_error("Incompatible domain data sample type");
+        }
+
+        const auto domainUnit = inputDomainDataDescriptor.getUnit();
+        if (domainUnit.getSymbol() != "s" && domainUnit.getSymbol() != "seconds")
+        {
+            throw std::runtime_error("Domain unit expected in seconds");
+        }
+
+        const auto domainRule = inputDomainDataDescriptor.getRule();
+        if (inputDomainDataDescriptor.getRule().getType() != DataRuleType::Linear)
+        {
+            throw std::runtime_error("FFT: Domain rule must be linear");
+        }
+
+        linearReader = BlockReaderFromExisting(linearReader, blockSize, SampleType::Float32, SampleType::UInt64);
+
+        auto dimensions = List<IDimension>();
+        dimensions.pushBack(Dimension(DimensionRuleBuilder()
+                                      .setType(DimensionRuleType::Linear)
+                                      .addParameter("delta", 1)
+                                      .addParameter("start", 1)
+                                      .addParameter("size", blockSize / 2)
+                                      .build()));
+
+        const auto inputRange = inputDataDescriptor.getValueRange();
+        outputDataDescriptor = DataDescriptorBuilder()
+                                   .setSampleType(SampleType::Float64)
+                                   .setDimensions(dimensions)
+                                   .setUnit(inputDataDescriptor.getUnit())
+                                   .setValueRange(inputRange.assigned() ? Range(0, inputRange.getHighValue()) : Range(0, 10))
+                                   .build();
+
+        outputSignal.setDescriptor(outputDataDescriptor);
+
+        outputDomainDataDescriptor =
+            DataDescriptorBuilderCopy(inputDomainDataDescriptor).setRule(ExplicitDataRule()).setSampleType(SampleType::UInt64).build();
+        outputDomainSignal.setDescriptor(outputDomainDataDescriptor);
+        configValid = true;
+    }
+    catch (const std::exception& e)
+    {
+        LOG_W("FFT: Failed to set descriptor for signal: {}", e.what())
+        outputSignal.setDescriptor(nullptr);
+    }
+}
+
+void FFTFbImpl::processEventPacket(const EventPacketPtr& packet)
+{
+    if (packet.getEventId() == event_packet_id::DATA_DESCRIPTOR_CHANGED)
+    {
+        const DataDescriptorPtr dataDesc = packet.getParameters().get(event_packet_param::DATA_DESCRIPTOR);
+        const DataDescriptorPtr domainDesc = packet.getParameters().get(event_packet_param::DOMAIN_DATA_DESCRIPTOR);
+        processSignalDescriptorChanged(dataDesc, domainDesc);
+    }
+}
+
+void FFTFbImpl::calculate()
+{
+    std::scoped_lock lock(sync);
+    SizeT readAmount = linearReader.getAvailableCount();
+    const auto inputData = std::make_unique<float[]>(readAmount * blockSize);
```

Comment:
Please move _all_ of the memory allocations and frees out of calculate(), probably into configure(). All of these block sizes can be calculated in advance, and consequently, all the buffers can be allocated in advance as well.

I am also pretty sure you don't need memset anywhere here, since you just override the buffers with actual values.

Reply:
True, both for the FFT input and the input data buffers we can pre-allocate. The FFT library is already a bit awkward as it has a struct type array, otherwise it could just get a memory ptr without any copies.